iroha_data_model/src/expression.rs

```diff
@@ -0,0 +1,838 @@
+    Vec(Vec<Value>),
+    /// `Id` of `Asset`, `Account`, etc.
+    Id(IdBox),
+    /// `Identifiable` as `Asset`, `Account` etc.
+    Identifiable(IdentifiableBox),
+    /// `PublicKey`.
+    PublicKey(PublicKey),
+}
+
+impl From<Value> for ExpressionBox {
+    fn from(value: Value) -> Self {
+        Expression::Raw(value.into()).into()
+    }
+}
+
+trait Evaluate {
+    fn evaluate(
+        &self,
+        world_state_view: &WorldStateView,
+        context: &Context,
+    ) -> Result<Value, String>;
+}
+
+impl Evaluate for Expression {
+    fn evaluate(
+        &self,
+        world_state_view: &WorldStateView,
+        context: &Context,
+    ) -> Result<Value, String> {
+        match self {
+            Expression::Add(add) => add.evaluate(world_state_view, context),
+            Expression::Subtract(subtract) => subtract.evaluate(world_state_view, context),
+            Expression::Greater(greater) => greater.evaluate(world_state_view, context),
+            Expression::Less(less) => less.evaluate(world_state_view, context),
+            Expression::Equal(equal) => equal.evaluate(world_state_view, context),
+            Expression::Not(not) => not.evaluate(world_state_view, context),
+            Expression::And(and) => and.evaluate(world_state_view, context),
+            Expression::Or(or) => or.evaluate(world_state_view, context),
+            Expression::If(if_expression) => if_expression.evaluate(world_state_view, context),
+            Expression::Raw(value) => Ok(*value.clone()),
+            Expression::Query => unimplemented!(),
+            Expression::Contains(contains) => contains.evaluate(world_state_view, context),
+            Expression::ContainsAll(contains_all) => {
+                contains_all.evaluate(world_state_view, context)
+            }
+            Expression::Where(where_expression) => {
+                where_expression.evaluate(world_state_view, context)
+            }
+            Expression::ContextValue(value_name) => context
+                .get(value_name)
+                .ok_or_else(|| format!("Value with name {} not found in context", value_name))
+                .map(|value| value.to_owned()),
+        }
+    }
+}
+
+/// Evaluates to the sum of right and left expressions.
+/// Works only for `Value::U32`
+#[derive(Debug, Clone)]
+pub struct Add {
+    /// Left operand.
+    pub left: ExpressionBox,
+    /// Right operand.
+    pub right: ExpressionBox,
+}
+
+impl Add {
+    /// Constructs `Add` expression.
+    pub fn new(left: ExpressionBox, right: ExpressionBox) -> Self {
+        Self { left, right }
+    }
+}
+
+impl Evaluate for Add {
+    fn evaluate(
+        &self,
+        world_state_view: &WorldStateView,
+        context: &Context,
+    ) -> Result<Value, String> {
+        let left = self.left.evaluate(world_state_view, context)?;
+        let right = self.right.evaluate(world_state_view, context)?;
+        if let (Value::U32(left), Value::U32(right)) = (left, right) {
```

Comment:
Why only U32?

Reply:
That's the only supported number in ISIs right now. I think we can generalize later.